File: leetcode/Trim a Binary Search Tree/2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {    
    
    TreeNode solve(TreeNode root, int low, int high){
        if(root == null) return null;

        if(root.val < low) return solve(root.right, low, high);
        if(root.val > high) return solve(root.left, low, high);

        root.right = solve(root.right, low, high);
        root.left = solve(root.left, low, high);
        return root;
    }
    
    public TreeNode trimBST(TreeNode root, int low, int high) {        
        return solve(root, low, high);
    }
}

File: leetcode/Trim a Binary Search Tree/2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func findReplacer(root *TreeNode, low, high int) *TreeNode {
	for root != nil && (root.Val > high || root.Val < low) {
		if root.Val > high {
			root = root.Left
		} else {
			root = root.Right
		}
	}
	return root
}

func trimBST(root *TreeNode, low int, high int) *TreeNode {
	root = findReplacer(root, low, high)
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		if curr == nil {
			continue
		}
		curr.Right = findReplacer(curr.Right, low, high)
		curr.Left = findReplacer(curr.Left, low, high)
		queue = append(queue, curr.Right, curr.Left)
	}

	return root
}